### ISSUE TITLE
Dark mode for plain text and "Unable to process type" message.

### ISSUE THREAD

**Issue by xiota** [xiota/geany-preview#2]
> I would like to change in the Preview-tab the white background of txt-documents.
>
> In those documents there is the text "Unable to process type" with white background. Could it be that a change of the background-color in Preview is not possible for those documents?

@ralf3u I'll look into it.

**Comment by xiota** [comment]
This should be fixed in 271f096.  It will be a little while before the PPA is updated.

The `extra_css` option now applies to the plain view.  There's an included `dark.css` and `invert.css` that you can customize, or you can write your own.  To refresh the css requires either restarting Geany or using the "Reload Config" button in *Edit/Plugin Preferences/Preview*.

**Comment by ralf3u** [comment]
@xiota

>>Now extra_css=disable shows a dark background (the background-color of sleepy pastel) in the sidebar of a txt-document. Is this the intention of the value extra_css=disable?

>The new css detects whether you are using a "dark" or "light" theme. You can look inside preview-headers.css to see the details.)


But if I look in preview.conf there is this text:
"disable*   - don't use any extra css files". So, maybe it would be better to write something like:
disable*   - detects automatically dark or light theme of the computer system (see result in preview-headers.css)

Before I continue testing, it would be good to change first the text, else I lose orientation.

question for a better understanding
In preview-headers.css I can see the color of sleepy pastel (background: #2e3436).
Did you code it, or did Geany Preview detected the dark system theme and then coded preview-headers.css automatically?
What if Geany Preview can't detect the system theme?

in preview.conf there is this text:
dark.css   - extra css file (use with the extra_css option)
Maybe it would be good to write something like this:
dark.css   - dark.css will be created automatically when extra_css=dark.css

in preview.conf there is this text:
invert.css   - extra css file
Maybe it would be good to write something like this:
invert.css   - invert.css will be created automatically when extra_css=invert.css

questions
What should be the background-color of txt-documents of dark.css?
What should be the background-color of html-documents of dark.css?
What should be the background-color of txt-documents of invert.css?
What should be the background-color of html-documents of invert.css?

If I use extra_css=disable: Why the background-color of html-documents is white?


Thank you for an answer in advance.

**Comment by ralf3u** [comment]
>It is used for email-style headers

I think it would be good to mention it in preview.css, at the top as comment, so users are informed.

**Comment by xiota** [comment]
@ralf3u  I found a workaround.  Try this:
```
<!DOCTYPE html>
<html>
<head><style></style></head>
<body>
<p>This is a test.</p>
</body>
</html>
```

If the document has another stylesheet, even if it is empty, the webview will use the rules in `extra-media.css`.  I've updated the plugin to insert an empty style if another one isn't already present.  The PPA has been updated.

**Comment by ralf3u** [comment]
>If the document has another stylesheet, even if it is empty, the webview will use the rules in extra-media.css.

That's cool! Now it works perfect for everyone's need.

But I still think that it would be good to mention in preview.css, at the top as comment, so users are informed, so something like: this document is used for the style of email headers.

**Comment by xiota** [comment]
> I still think that it would be good to mention in preview.css, at the top as comment

It is mentioned in the config file, and the name of the rules say "header" in them.  I will add more details to the questions document later.

**Comment by ralf3u** [comment]
>It is mentioned in the config file

I see it now. I forgot to scroll upwards. But why there is in the comments an empty line before and after preview.css?

**Comment by ralf3u** [comment]
I just updated the system.


issue 1
There is still in preview.conf an empty line before and after preview.css, although I think it is not necessary.

issue 2
In geany.css the background of geany-preview-sidebar-page is not respected any more.

issue 3
In the code in comment https://github.com/xiota/geany-preview/issues/2#issuecomment-956280646 I think you forgot to mention geany-xitweaks-sidebar-tab-focus.

issue 4
Now there are no small white points around the tab-title Preview. But the problem is now: There is a colored border around the content of the tabs of the sidebar in different colors. So I deleted geany-xitweaks-sidebar-page-focus and geany-xitweaks-sidebar-page-unfocus. And now there is not a colored border any more around the content of the tabs of the sidebar in different colors.

**Comment by xiota** [comment]
> There is still in preview.conf an empty line before and after preview.css, although I think it is not necessary.

Preview has not been updated.  I'm going to wait for more substantial coding changes before updating it.

>  In geany.css the background of geany-preview-sidebar-page is not respected any more.

Try adding the background to the `#geany-xitweaks-sidebar-page-unfocus`.  Styling the page when it has focus messes up other styling.

**Comment by xiota** [comment]
The following rule gets rid of the outline:
```
notebook header tabs > * {
	outline-width: 0;
}
```
The following can be used to style the tabs and pages of the sidebar:
```
#geany-xitweaks-sidebar-tab-focus {
}
#geany-xitweaks-sidebar-tab-unfocus {
}
#geany-xitweaks-sidebar-page-focus {
}
#geany-xitweaks-sidebar-page-unfocus {
}
```
The following is for the preview page only, but it gets over-written when the focus feature is enabled.
```
#geany-preview-sidebar-page {
}
```

**Comment by ralf3u** [comment]
>You had previously stated: "The focus should not be on the tab-title but on the content of the tab."

I think there is a misunderstanding. The citation is inserted out of the context.
What I wrote was this:
>possible solution
The focus should not be on the tab-title but on the content of the tab.

The original intention was to avoid the white points. And for that a possible solution could be to put the focus on the content.

**Comment by xiota** [comment]
@ralf3u  Whatever the issue, you should now be able to style the tabs *and* pages however you like.  If you don't want to style the pages, just leave those rules out.

**Comment by xiota** [comment]
> Does that mean that I can use either geany-xitweaks-sidebar-tab-focus or geany-preview-sidebar-page, but not both at the same time?

Doesn't the background of the preview page only need to be active when Geany is first started?  Can you set the unfocused style to match the preview page style?

**Comment by ralf3u** [comment]
>The following is for the preview page only, but it gets over-written when the focus feature is enabled.

>Doesn't the background of the preview page only need to be active when Geany is first started? Can you set the unfocused style to match the preview page style?

Did I understand it right:

question 1
Can use either geany-xitweaks-sidebar-tab-focus or geany-preview-sidebar-page, but not both at the same time?

question 2
So, instead of using geany-xitweaks-sidebar-tab-focus and geany-preview-sidebar-page, because it gets over-written, it is better to use geany-xitweaks-sidebar-tab-focus and geany-xitweaks-sidebar-page-unfocus?

**Comment by xiota** [comment]
> So, instead of using geany-xitweaks-sidebar-tab-focus and geany-preview-sidebar-page, because the last one gets over-written, it is better to use geany-xitweaks-sidebar-tab-focus and geany-xitweaks-sidebar-page-unfocus?

On my computer, it works fine.  There may be a race condition on your computer.  If it works on your computer with them both set, I'd rather leave it alone since this feature was very difficult to the point it is now.

**Comment by xiota** [comment]
> On this webpage... the code should be changed...

The page mentions the `#geany-preview-sidebar-page` rule because that is the one implemented in the preview plugin.  The other one goes with a different plugin.

> Now, I can't find an issue any more. It's perfect.

Glad it's *finally* working.  Thank you for testing the plugin and providing feedback.

**Comment by ralf3u** [comment]
steps to reproduce:
1 open an html-document
2 editor is in focus
3 with the mouse: mark text in the Preview-tab

result: "Preview" is not green
expectation: "Preview" should be green

**Comment by xiota** [comment]
I've already looked through all the signals a few times searching for signals that will work for this feature.  I don't know what else to try.

**Comment by ralf3u** [comment]
>I've already looked through all the signals a few times searching for signals that will work for this feature. I don't know what else to try.

So, unfortunately it seems to be that issue number 5 ("when sidebar is in focus: visible effect of the tab") can't be fixed completely.

issue a
In xittweaks.conf there is menubar_hide_on_start=true.
But if in Geany I hide the menubar and I restart Geany, then the menubar is visible.

issue b
By pressing the shortcut, now the tab-title is in focus. Days before the white points appeared for the first time, I'm sure that the content was in focus. Some days ago I wrote here on this webpage that for me it does not matter if tab-title or the content is in focus. But I realized now that it is much better if the content is in focus when the shortcut toggle editor/sidebar is pressed. Could you please change it like before, so that when the shortcut toggle editor/sidebar is pressed that not the tab-title of the sidebar is in focus but the content of the sidebar?

issue c
I don't know what happened, but since yesterday if I press the shortcut toggle editor/sidebar, "Preview" is not green any more. Same when I click on the tabs of the sidebar: The tab-titles are not green any more, although in xittweaks.conf there is sidebar_focus_enabled=true. And also the white points are back in the tab-titles of the sidebar, although in geany.css there is the same code like https://github.com/xiota/geany-preview/issues/2#issuecomment-958176069. And with the same code I can see again the gray background when restarting Geany. And this can't be fixed by replacing the code with this code:
```
#geany-preview-sidebar-page {
	background: white;
}
```

Thank you for an answer in advance.

**Comment by xiota** [comment]
> But if in Geany I hide the menubar and I restart Geany, then the menubar is visible.

I've updated the PPA.  After installing, please reset the config file, and re-enter your preferences from scratch.

Also, when you're done editing the config, reload the settings from the menu.  Otherwise, the plugin will overwrite your changes with the old settings when Geany shuts down.  (I'll move the settings that need to be saved on shutdown to a different file in a future version.)

> Could you please change it like before, so that when the shortcut toggle editor/sidebar is pressed that not the tab-title "Preview" is in focus but the content of the Preview-tab?

If I understand what you're asking, you can do it by clearing out the old rules in `geany.css`.  Then try the following:
```CSS
notebook header tabs > * {
	outline-width: 0;
}

#geany-xitweaks-sidebar-tab-focus {
	font-weight: bold;
	color: #46af41;
}
#geany-xitweaks-sidebar-tab-unfocus {
	/* Whatever you want */
}

#geany-xitweaks-sidebar-page-focus {
	border-width: 3px;
	border-style: solid;
	border-color: #46af41;
}

#geany-xitweaks-sidebar-page-unfocus {
	border-width: 3px;
	border-style: solid;
	border-color: transparent;
	background: #2e3436;
}

#geany-preview-sidebar-page {
	border-width: 3px;
	border-style: solid;
	border-color: transparent;
	background: #2e3436;
}
```

**Comment by ralf3u** [comment]
I did an update. I deleted the config-file of preview and I delted the config-file of xitweaks, and I deleted all css-files of preview, so I don't need to reload them. And I deleted geany.css. Then I restarted Geany.

issue a was fixed

issue b was not fixed.
If I press the shortcut toggle editor/sidebar, then the tab-title "Preview" is in focus.
The goal should be: Not the tab-title "Preview" should be in focus, but the content of the Preview-tab.
Two tests if the goal was achieved with a very long html-document, and with at least two tabs in the sidebar:
test 1:
Open Geany, press the shortcut toggle editor/sidebar, then press the arrow-left-key once: the Preview-tab should still be in focus.
test 2:
Open Geany, press the shortcut toggle editor/sidebar, then press the arrow-down-key once: the html-document should move downwards.

issue c
I added in geany.css the rules, mentioned in the previous comment
- no green color, like before
- there are no white points in the tab-title any more, so this has changed in a positive way
- background: when I open Geany with an html-document there is a white flash now in the sidebar, and I don't know why; extra_css=disable, and in the html-document, in the body-tag there is this in the style: background:#2e3436; so, there should be no white flash

**Comment by xiota** [comment]
In `geany.css`, these rules control the tab style.
```CSS
#geany-xitweaks-sidebar-tab-focus {
}
#geany-xitweaks-sidebar-tab-unfocus {
}
```
These rules control the page:
```CSS
#geany-xitweaks-sidebar-page-focus {
}
#geany-xitweaks-sidebar-page-unfocus {
}
#geany-preview-sidebar-page {
}
```
This rule controls the white outline on the tab:
```CSS
notebook header tabs > * {
	outline-width: 0;
}
```

With `extra_css=disable`, the default webview color, before the document loads, is white.  Set it to something to change the default color.  The `geany.css` style only controls the window style, which is different from the webview style.

**Comment by ralf3u** [comment]
>The geany.css style only controls the window style, which is different from the webview style.

OK, background is now fixed. There is no white flash any more.

And what about the issue b?
And what about the green color that is missing?

Thank you for an answer in advance.

**Comment by xiota** [comment]
> The goal should be: Not the tab-title "Preview" should be in focus, but the content of the Preview-tab.

You need to edit `geany.css` to make it do what you want.  I don't know what rules you are using.

**Comment by ralf3u** [comment]
>You need to edit geany.css to make it do what you want. I don't know what rules you are using.

isuue b: I'm sure that a solution of issue b has nothing to do with geany.css.
issue c: the absence of the green color: I use the values of https://github.com/xiota/geany-preview/issues/2#issuecomment-961166111.

**Comment by xiota** [comment]
I don't know what the issues are now.  Please re-describe them.  Don't link back or copy from a previous comment because I do not understand your previous descriptions.

For now, make `geany.css` have the [contents from earlier](https://github.com/xiota/geany-preview/issues/2#issuecomment-961166111), even if it doesn't do exactly what you want (tab text color and bold) so I can figure out if there is another issue.

**Comment by ralf3u** [comment]
>For now, make geany.css have the contents from earlier, even if it doesn't do exactly what you want (tab text color and bold) so I can figure out if there is another issue.

<br>

**issue shortcut**
If I press the shortcut toggle editor/sidebar, then the tab-title "Preview" is in focus.
The goal should be: Not the tab-title "Preview" should be in focus, but the content of the Preview-tab.
Two tests if the goal was achieved with a very long html-document, and with at least two tabs in the sidebar:
TEST 1: Open Geany, press the shortcut toggle editor/sidebar, then press the arrow-left-key once: the Preview-tab should still be in focus.
TEST 2: Open Geany, press the shortcut toggle editor/sidebar, then press the arrow-down-key once: the html-document should move downwards.

<br>

**issue green color**
- the absence of the green color
- something is strange: in an html-document, an a-tag without style: when I go in the Preview-tab with the mouse over the link, the link becomes green, although it should not become green (it will not become green in the browsers Firefox and Web)

**Comment by xiota** [comment]
Thanks for explaining.  If it seems that I don't understand again, it might help to try another explanation.

> If I press the shortcut toggle editor/sidebar, then the tab-title "Preview" is in focus.
> The goal should be: Not the tab-title "Preview" should be in focus, but the content of the Preview-tab.

When you say "focus" are you referring to the color or the part of the widget that takes keyboard input?

> issue green color
>  "Preview" is not green when I click on it or when I press the shortcut toggle editor/sidebar

There are various issues with this feature that never really worked.  Is this one of those?  In the future when referencing these types of problems, please note that they are "old issue, still not fixed".  That way I don't get confused and think you're trying to describe something new.

> something is strange: in an html-document, an a-tag without style: when I go in the Preview-tab with the mouse over the link, the link becomes green, although it should not become green (it will not become green in the browsers Firefox and Web)

Are you using the `extra_css` option?  The `extra-media.css` file does style links.

**Comment by ralf3u** [comment]
>When you say "focus" are you referring to the color or the part of the widget that takes keyboard input?

No, there is no color change. I think that it is in focus because if if I do both test then it seems to be that it is in focus. So here the changed report:

If I press the shortcut toggle editor/sidebar, then the tab-title "Preview" seems to be in focus.
But the goal should be: Not the tab-title "Preview" should be in focus, but the content of the Preview-tab.

control tests with a very long html-document, and with at least two tabs in the sidebar:
TEST 1: Open Geany, press the shortcut toggle editor/sidebar, then press the arrow-left-key only once: the content of the Preview-tab should still be in focus.
TEST 2: Open Geany, press the shortcut toggle editor/sidebar, then press the arrow-down-key only once: the html-document should move downwards.

**Comment by ralf3u** [comment]
>issue green color
"Preview" is not green when I click on it or when I press the shortcut toggle editor/sidebar

feedback
Yesterday morning I did an update: The green color is back.
This morning I did an update: The green color is back.

**Comment by xiota** [comment]
> This morning I did an update: The green color is back.

I don't know if the problem is whether you want the green color or not.

I don't know what is becoming green.  If it is the tab, label, that is controlled by `geany.css`.  If it is a link within the webpage, it is controlled by the `css` files inside `.config/geany/plugins/preview/`.

**Comment by ralf3u** [comment]
>I don't know if the problem is whether you want the green color or not.

Now where the green color is back, it is not a problem any more. Some days ago the green color was not there any more. With green color I mean the tab-title in the sidebar.

**Comment by xiota** [comment]
> it is not a problem any more

Thanks for letting me know.  I don't know if I'll be able to get this feature working properly because GTK doesn't seem to have the signals needed to reliably detect when the widget has focus.  I'll keep it in mind in case I come across a solution.

**Comment by ralf3u** [comment]
To the green color: I would like to make a better differentiation, so there is no misunderstanding:
- shortcut toggle editor/sidebar is pressed: Preview is green, so it works
- clicking in the sidebar on a tab-title: tab-title is green, so it works
- clicking in the sidebar in the content of a tab: tab-title is not green, so it does not work (old issue, still not fixed)

**Comment by xiota** [comment]
@ralf3u Thank you for the update/clarification.
